### PR TITLE
Ignore actual members in UnusedPrivateMember

### PR DESCRIPTION
**Pull request by sczerwinski**
The general rule of thumb should be:
- if a rule is ignored for `abstract` members, it should be ignored for `expect` members,
- if a rule is ignored for **overridden** members, it should be ignored for `actual` members.

See also:
- #3388
- #3390

Related to #3415

### CODE REVIEW

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt [hunk 1]
```diff
@@ -204,7 +205,7 @@ private class UnusedParameterVisitor(allowedNames: Regex) : UnusedMemberVisitor(
     }
 
     override fun visitClassOrObject(klassOrObject: KtClassOrObject) {
-        if (klassOrObject.isExpect()) return
+        if (klassOrObject.isExpect() || klassOrObject.isActual()) return
```
Comment (chao2zhang) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt:208]: How is this related to the rule of thumb `if a rule is ignored for overridden members, it should be ignored for actual members.`?

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt [hunk 2]
```diff
@@ -292,7 +294,11 @@ private class UnusedPropertyVisitor(allowedNames: Regex) : UnusedMemberVisitor(a
     override fun visitPrimaryConstructor(constructor: KtPrimaryConstructor) {
         super.visitPrimaryConstructor(constructor)
         constructor.valueParameters
-            .filter { (it.isPrivate() || !it.hasValOrVar()) && it.containingClassOrObject?.isExpect() == false }
+            .filter {
+                (it.isPrivate() || !it.hasValOrVar()) &&
+                    it.containingClassOrObject?.isExpect() == false &&
+                    it.containingClassOrObject?.isActual() == false
```
Comment (chao2zhang) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt:300]: How is this change related to the rule of thumb `if a rule is ignored for overridden members, it should be ignored for actual members.`?

Comment (sczerwinski) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt:300]: There is no such thing as abstract or overridden constructors, but there can be `expect` or `actual` constructors, and those should be treated the same way as `expect`/`actual` functions (line 259-260).

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt (hunk 1 above)
Comment (sczerwinski) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt:208]: This line is actually my mistake (I went with `isExpect()` one step too far). Actual classes or objects should be visited. Just their actual members should be ignored.

I'll update the PR.

**Comment by sczerwinski** [comment]
@chao2zhang, Sorry for the mistake. I've fixed the PR, rebased and squashed for better readability.

Also an example of an actual constructor, that should be ignored:
**commonMain:**
```kotlin
expect class Foo(bar: Bar) {
    expect fun doSomething()
}
```
**jvmMain:**
```kotlin
actual class Foo actual constructor(private val bar: Bar) {
    actual fun doSomething() {
        bar.doSomething()
    }
}
```
**jsMain:**
```kotlin
actual class Foo actual constructor(bar: Bar) {
    actual fun doSomething() {
        // Do something with JS that doesn't require bar
    }
}
```

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt (hunk 2 above)
Comment (sczerwinski) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt]: I added more test cases and changed this snippet. My initial condition was excluding too many cases.

Comment (chao2zhang) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt]: Thanks for the detailed explanation and additional test cases.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt [hunk 3]
```diff
@@ -292,7 +294,10 @@ private class UnusedPropertyVisitor(allowedNames: Regex) : UnusedMemberVisitor(a
     override fun visitPrimaryConstructor(constructor: KtPrimaryConstructor) {
         super.visitPrimaryConstructor(constructor)
         constructor.valueParameters
-            .filter { (it.isPrivate() || !it.hasValOrVar()) && it.containingClassOrObject?.isExpect() == false }
+            .filter {
+                (it.isPrivate() || (!it.hasValOrVar() && !constructor.isActual())) &&
+                    it.containingClassOrObject?.isExpect() == false
```
Comment (chao2zhang) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt:299]: I am actually planning to separate UnusedPrivateParameter from UnusedPrivateMember, so the logic here should be simplified.

**Review by chao2zhang — APPROVED** [review]

**Review by BraisGabin — APPROVED** [review]
